File: Java/BFS/RightView.java
/*
*   @author
*   Aakash Verma
*
*   Output:
*   12 1 5 3
*
*/

import java.util.*;

class Node {
    int data;
    Node left, right;
    public Node(int key) {
        data = key;
        left = right = null;
    }
}


class RightView {

    static Node root;
    RightView() {
        root = null;
    }

    public static List<Integer> rightView(Node root) {
        List<Integer> result = new ArrayList<>();
        if(root == null) return result;
        Queue<Node> queue = new LinkedList<>();
        queue.offer(root);
        while(!queue.isEmpty()) {
            int levelSize = queue.size();
            for(int i = 0; i < levelSize; i++) {
                Node current = queue.poll();
                
                if(i == levelSize - 1) {
                    result.add(current.data);
                }

                if(current.left != null) 
                    queue.offer(current.left);
                if(current.right != null)
                    queue.offer(current.right);
            }
        }
        return result;
    }

    public static void main (String[] args) {
        root = new Node(12);
        root.left = new Node(7);
        root.right = new Node(1);
        root.left.left = new Node(9);
        root.right.left = new Node(10);
        root.right.right = new Node(5);
        root.left.left.left = new Node(3);
        List<Integer> ans = rightView(root);
        for(int node : ans) {
            System.out.print(node + " ");
        }
        System.out.println();
    }
}

File: Python/BFS/RightView.py
# @author
# Aakash Verma

# Output:
# 12 1 5 3

from collections import *

class Node:
    def __init__(self, data):
        self.data = data
        self.right = self.left = None

class RightView:
    
    def __init__(self): 
        self.root = None

    def rightView(self, root):
        result  = []
        if root is None:
            return result
        queue = deque([])
        queue.append(self.root)
        while len(queue) != 0:
            level_size = len(queue)
            for i in range(level_size):
                current = queue.popleft()
                
                if (i == level_size - 1):
                    result.append(current.data)
                
                if current.left is not None:
                    queue.append(current.left)
                if current.right is not None:
                    queue.append(current.right)
        return result
    
if __name__ == '__main__':
    tree = RightView()
    tree.root = Node(12);
    tree.root.left = Node(7);
    tree.root.right = Node(1);
    tree.root.left.left = Node(9);
    tree.root.right.left = Node(10);
    tree.root.right.right = Node(5);
    tree.root.left.left.left = Node(3);
    ans = tree.rightView(tree.root)
    for node in ans:
        print(node, end = " ")
    print()


